In [5]:
import pandas as pd
import numpy as np
import os, fnmatch

In [6]:
root = '/Users/Lee/Documents/jupyter/notebooks/CP207/CPS Data'
#Get all of the cost reports
matches = []
for root, dirnames, filenames in os.walk(root):
  for filename in fnmatch.filter(filenames, '*.dta'):
    matches.append(os.path.join(root, filename))

In [8]:
len(matches)

16

In [ ]:
df_soft_costs = pd.DataFrame()
for item in matches[0:]:
    if not df_soft_costs.empty:
        df_soft_costs = df_soft_costs.append(soft_costs_processed(item, verbose=False))
    else:
        df_soft_costs = soft_costs_processed(item, verbose=False)

In [9]:
df = pd.read_stata('CPS Data/2015/cpsmar2015.dta', convert_categoricals = False)

In [10]:
df.head()

hrecord  h_seq  hhpos  hunits  hefaminc  h_respnm  h_year  h_hhtype  \
0        1      2      0       1         9         2    2015         1   
1        1      2      0       1         9         2    2015         1   
2        1      2      0       1         9         2    2015         1   
3        1      2      0       1         9         2    2015         1   
4        1      3      0       5         4         1    2015         1   

   h_numper  hnumfam   ...    m5gsame  m5g_reg  m5g_st  m5g_div  m5g_mtr1  \
0         4        2   ...          1        1       0        1         1   
1         4        2   ...          1        1       0        1         1   
2         4        2   ...          1        1       0        1         1   
3         4        2   ...          0        0       0        0         8   
4         1        1   ...          1        1       0        1         1   

   m5g_mtr3  m5g_mtr4  i_m5g1  i_m5g2  i_m5g3  
0         1         1       5       0       0  
1         1         1       1       0       0  
2         1         1       5       0       0  
3         8         9       0       0       0  
4         1         1       0       0       0  

[5 rows x 704 columns]

In [11]:
columns_2015 = df.columns.tolist()

In [12]:
[column for column in columns_2015 if 'rp' in column]

['hrpaidcc', 'a_exprrp', 'perrp', 'prpertyp', 'prptrea', 'axrrp']

In [13]:
[column for column in columns_2015 if 'age' in column]

['a_age', 'age1', 'wageotr', 'axage']

In [14]:
[column for column in columns_2015 if 'cbsa' in column]

['gtcbsa', 'gtcbsast', 'gtcbsasz']

In [15]:
df[df.a_age < 24].head()

hrecord  h_seq  hhpos  hunits  hefaminc  h_respnm  h_year  h_hhtype  \
2         1      2      0       1         9         2    2015         1   
3         1      2      0       1         9         2    2015         1   
6         1      5      0       1         7         1    2015         1   
19        1     13      0       2         6         2    2015         1   
20        1     13      0       2         6         2    2015         1   

    h_numper  hnumfam   ...    m5gsame  m5g_reg  m5g_st  m5g_div  m5g_mtr1  \
2          4        2   ...          1        1       0        1         1   
3          4        2   ...          0        0       0        0         8   
6          2        1   ...          1        1       0        1         1   
19         5        1   ...          0        0       0        0         8   
20         5        1   ...          0        0       0        0         8   

    m5g_mtr3  m5g_mtr4  i_m5g1  i_m5g2  i_m5g3  
2          1         1       5       0       0  
3          8         9       0       0       0  
6          1         1       0       0       0  
19         8         9       0       0       0  
20         8         9       0       0       0  

[5 rows x 704 columns]

San Francisco-Oakland-Hayward CBSA = 41860

In [19]:
bay_households = df[df.gtcbsa == 41860][['h_seq', 'h_numper']]

In [20]:
bay_households = bay_households.drop_duplicates()

In [21]:
len(bay_households)

805

In [22]:
df[df.gtcbsa == 41860].head()

hrecord  h_seq  hhpos  hunits  hefaminc  h_respnm  h_year  h_hhtype  \
188353        1  94631      0       5         8         1    2015         1   
188354        1  94631      0       5         8         1    2015         1   
188355        1  94631      0       5         8         1    2015         1   
188356        1  94631      0       5         8         1    2015         1   
188357        1  94631      0       5         8         1    2015         1   

        h_numper  hnumfam   ...    m5gsame  m5g_reg  m5g_st  m5g_div  \
188353         7        1   ...          2        4       6        9   
188354         7        1   ...          2        4       6        9   
188355         7        1   ...          2        4       6        9   
188356         7        1   ...          2        4       6        9   
188357         7        1   ...          2        4       6        9   

        m5g_mtr1  m5g_mtr3  m5g_mtr4  i_m5g1  i_m5g2  i_m5g3  
188353         2         2         2       0       0       0  
188354         2         2         2       0       0       0  
188355         2         2         2       0       0       0  
188356         2         2         2       0       0       0  
188357         2         2         2       0       0       0  

[5 rows x 704 columns]

In [23]:
len(df.a_age.unique())

82

In [16]:
df.age1.unique()

array([10,  5,  0, 17, 14,  2, 16, 15, 13,  6, 12,  8,  9, 11,  7,  4,  1,
        3])

In [36]:
df2 = pd.read_stata('CPS Data/2014/cpsmar2014.dta', convert_categoricals = False)

In [8]:
df2.head()

hrecord  h_seq  hhpos  hunits  hefaminc  h_respnm  h_year  h_hhtype  \
0        1      3      0       1         9         2    2014         1   
1        1      3      0       1         9         2    2014         1   
2        1      4      0       1         7         1    2014         1   
3        1      4      0       1         7         1    2014         1   
4        1      6      0       1        12         1    2014         1   

   h_numper  hnumfam    ...      tdiv_val  trnt_val  ted_val  tcsp_val  \
0         2        1    ...             0         0        0         0   
1         2        1    ...             0         0        0         0   
2         2        1    ...             0         0        0         0   
3         2        1    ...             0         0        0         0   
4         2        1    ...             0         0        0         0   

   tfin_val  toi_val  tphip_val  tpotc_val  tpmed_val  tchsp_val  
0         0        0          0          0          0          0  
1         0        0          0          0          0          0  
2         0        0          0          0          0          0  
3         0        0          0          0          0          0  
4         0        0          0          0          0          0  

[5 rows x 700 columns]

In [9]:
print len(df)
print len(df2)

199024
60141


In [10]:
df = df.append(df2)

In [11]:
len(df)

259165

In [12]:
df.head()

a_age  a_civlf  a_clswkr  a_dtind  a_dtocc  a_enrlw  a_ernlwt  a_explf  \
0     48        1         1       23       22        2         0        1   
1     45        0         0        0        0        2         0        0   
2     23        1         1       22       16        2         0        1   
3      2        0         0        0        0        0         0        0   
4     80        0         0        0        0        0         0        0   

   a_exprrp  a_famnum  ...    wicyn  wicyna  wkcheck  wkswork  workyn  wrk_ck  \
0         1         1  ...        0       0        3       52       1       1   
1         4         1  ...        2       1        1       17       1       1   
2        12         2  ...        2       1        3       52       1       1   
3        12         2  ...        2       1        0        0       0       0   
4         2         0  ...        0       0        0        0       2       2   

   ws_val  wsal_val  wsal_yn  wtemp  
0       0     40000        1      0  
1       0      6000        1      0  
2       0     16000        1      0  
3       0         0        0      0  
4       0         0        2      2  

[5 rows x 712 columns]

"Living independently means heading one's own household or living in a household headed by a spouse, unmarried partner, or other non-relative. Young adults are 18- to 34-year olds, excludes 18- to 24-year-old college students enrolled full time"

So we need to identify the following fields/people:
-Age
-Household
-Head of household
-College enrollment status

Also need to identify how these have changed across time - Pew study looked at 2007 (or 2000?) onward.

"'Doubled up' mean living with one or more adults who is not your spouse or unmarried partner; those living with parents or other relatives are considered 'doubled up.' Young adults are 18- to 34-year olds, excludes 18- to 24-year-old college students enrolled full time"

In [39]:
df.columns[0:10].tolist()+['h_seq','parent']

['hrecord',
 'h_seq',
 'hhpos',
 'hunits',
 'hefaminc',
 'h_respnm',
 'h_year',
 'h_hhtype',
 'h_numper',
 'hnumfam',
 'h_seq',
 'parent']

In [41]:
df.parent.unique()

array([0, 2, 4, 1, 3])

In [ ]:
df[df.columns[0:10].tolist()+['h_seq','parent']].head()

In [24]:
[column for column in columns_2015 if 'line' in column]

['a_lineno']

In [25]:
df.a_famrel.value_counts()

3    66243
1    55197
2    38963
0    31165
4     7456
Name: a_famrel, dtype: int64

In [27]:
df.a_hscol.value_counts()

0    180923
2     10062
1      8039
Name: a_hscol, dtype: int64

In [28]:
df.a_ftpt.value_counts()

0    180923
1     15516
2      2585
Name: a_ftpt, dtype: int64

In [15]:
len(df)

199024

### Variables deciphered so far:

- h_seq: is the same across a household
- a_age: indicates age
- perrp: indicates "Relationship to reference" (need to investigate more)
- a_fam_rel: indicates "family relation" (perhaps better than perrp)
- prdtrace: race
- a_lineno: line number of the person (used for things like "line number of parent")
- a_ftpt: full time or part time in school
- a_hscol: high school or college
- h_type: household type (married/unmarried, other types)

Need to identify:
PESCHFT (Are you enrolled in school as a full-time or part-time student?)
PESCHLVL (High school, college, or university?)
HRHTYPE (Household type, may help us decipher family or not)

"Living independently means heading one's own household or living in a household headed by a spouse, unmarried partner, or other non-relative. Young adults are 18- to 34-year olds, excludes 18- to 24-year-old college students enrolled full time"

So we need to identify the following fields/people:
-Age
-Household
-Head of household
-College enrollment status

Also need to identify how these have changed across time - Pew study looked at 2007 (or 2000?) onward.

In [30]:
def in_sample(row):
    #In sample if between 18 and 34 AND (not in college or not full time)
    if row['a_age'] >= 18 and row['a_age'] < 35 and not (row['a_age'] < 25 and (row['a_hscol'] == 2 and row['a_ftpt'] == 1)):
        return 1
    else:
        return 0

In [31]:
df['Millenial'] = df.apply(lambda row: 1 if row['a_age'] >= 18 and row['a_age'] < 35 else 0, axis = 1)

In [32]:
df['FT College'] = df.apply(lambda row: 1 if row['a_hscol'] == 2 and row['a_ftpt'] == 1 else 0, axis = 1)

In [33]:
pd.pivot_table(df, values='h_seq', index=['Millenial'], columns=['FT College'], aggfunc=np.count_nonzero)

FT College       0     1
Millenial               
0           155420   821
1            35953  6830

So there are 821 non-millenials who are FT college students

Old method without college status

In [ ]:
#df['In Sample'] = df.apply(lambda x: 1 if x['a_age'] >= 18 and x['a_age'] < 35 else 0, axis = 1)

In [26]:
#df['In Sample'].value_counts()

0    156241
1     42783
dtype: int64

In [34]:
pd.pivot_table(df, values='h_seq', index=['a_hscol'], columns=['a_ftpt'], aggfunc=np.count_nonzero)

a_ftpt        0     1     2
a_hscol                    
0        180923   NaN   NaN
1           NaN  7865   174
2           NaN  7651  2411

In [40]:
42783-7651

35132

In [35]:
df['In Sample'] = df.apply(in_sample, axis = 1)

In [36]:
pd.pivot_table(df, values='h_seq', index=['In Sample'], columns=['a_ftpt'], aggfunc=np.count_nonzero)

a_ftpt          0      1     2
In Sample                     
0          148812  11908   823
1           32111   3608  1762

In [37]:
df['In Sample'].value_counts()

0    161543
1     37481
Name: In Sample, dtype: int64

In [38]:
df['h_type'].value_counts()

1    123153
4     32814
7     14138
6     13949
3     12711
2      1971
5       119
9       115
8        54
Name: h_type, dtype: int64

- 0   NON-INTERVIEW HOUSEHOLD
- 1	HUSBAND/WIFE PRIMARY FAMILY (NEITHER AF)
- 2	HUSB/WIFE PRIM. FAMILY (EITHER/BOTH AF)
- 3	UNMARRIED CIVILIAN MALE-PRIM. FAM HHLDER
- 4	UNMARRIED CIV. FEMALE-PRIM FAM HHLDER
- 5	PRIMARY FAMILY HHLDER-RP IN AF, UNMAR.
- 6	CIVILIAN MALE PRIMARY INDIVIDUAL
- 7	CIVILIAN FEMALE PRIMARY INDIVIDUAL
- 8	PRIMARY INDIVIDUAL HHLD-RP IN AF
- 9	GROUP QUARTERS WITH FAMILY
- 10	GROUP QUARTERS WITHOUT FAMILY

AF = Armed forces
RP = Related person?

In [39]:
df['perrp'].value_counts()

4     64008
1     51714
3     37469
2     22543
14     4355
5      4258
8      3686
6      3257
7      2312
16     2237
12      977
10      773
18      465
13      352
15      346
9       192
17       80
Name: perrp, dtype: int64

- 1	REFERENCE PERSON W/RELS.
- 2	REFERENCE PERSON W/O RELS.
- 3	SPOUSE
- 4	CHILD
- 5	GRANDCHILD
- 6	PARENT
- 7	BROTHER/SISTER
- 8	OTHER REL. OR REF. PERSON
- 9	FOSTER CHILD
- 10	NONREL. OF REF. PERSON W/RELS.
- 11	NOT USED
- 12	NONREL. OF REF. PERSON W/O RELS.
- 13	UNMARRIED PARTNER W/RELS.
- 14	UNMARRIED PARTNER W/OUT RELS.
- 15	HOUSEMATE/ROOMMATE W/RELS.
- 16	HOUSEMATE/ROOMMATE W/OUT RELS.
- 17	ROOMER/BOARDER W/RELS.
- 18	ROOMER/BOARDER W/OUT RELS.

In [18]:
df[df.h_seq == 10][['a_lineno', 'perrp', 'a_age']]

a_lineno  perrp  a_age
13         1      1     63
14         2      3     61
15         3      4     26

In [40]:
def independently(row):
    if row['perrp'] in [1, 2, 3, 10, 12, 13, 14, 15, 16, 17 , 18]:
        return 1
    else:
        return 0

In [41]:
df['Living Independently'] = df.apply(independently, axis = 1)

In [42]:
pd.pivot_table(df, values='h_seq', index=['Living Independently'], columns=['In Sample'], aggfunc=np.count_nonzero)

In Sample                 0      1
Living Independently              
0                     65737  11976
1                     95806  25505

In [27]:
df[df['In Sample'] == 1]['perrp'].value_counts()

1     10312
4      9516
3      6421
2      4463
14     2054
16     1211
8       987
7       841
5       583
12      390
18      217
13      149
10      140
15      122
6        26
17       26
9        23
dtype: int64

heading one's own household or living in a household headed by a spouse, unmarried partner, or other non-relative.

In [43]:
#Testing if parents are living with people who are considered head of household
df[(df['In Sample'] == 1) & (df['perrp'] == 6)][['a_lineno', 'perrp', 'a_age', 'h_seq']].head(10)

a_lineno  perrp  a_age  h_seq
13468         3      6     30   8469
17080         2      6     30   9656
22873         2      6     33  12582
40150         2      6     32  21098
55582         2      6     34  28862
59464         2      6     32  30592
66382         2      6     30  34812
70657         3      6     29  36483
71348         2      6     29  36659
85755         3      6     31  44696

In [33]:
df[df['h_seq'] == 12582][['a_lineno', 'perrp', 'a_age', 'h_seq']]

a_lineno  perrp  a_age  h_seq
22872         1      1     16  12582
22873         2      6     33  12582
22874         3      6     38  12582

In [37]:
df2['In Sample'] = df2.apply(in_sample, axis = 1)

In [38]:
df2['Living Independently'] = df2.apply(independently, axis = 1)